File: draft_design_pattern/interpreter.nim
type
  IValue = ref object of RootObj
  Value = ref object of IValue
    value: int
  Add = ref object of IValue
    value1: IValue
    value2: IValue
  Mul = ref object of IValue
    value1: IValue
    value2: IValue


method getValue(self: IValue): int {.base, locks: "unknown".} =
  discard


method getValue(self: Value): int {.locks: "unknown".} =
  self.value


method getValue(self: Add): int {.locks: "unknown".} =
  self.value1.getValue() + self.value2.getValue()


method getValue(self: Mul): int {.locks: "unknown".} =
  self.value1.getValue() * self.value2.getValue()


let obj = Mul(
  value1: Add(value1: Value(value: 1), value2: Value(value: 2)),
  value2: Value(value: 3)
)

echo obj.getValue()

File: draft_design_pattern/interpreter.ts
export {};

interface IValue {
  getValue(): number;
}

class Value implements IValue {
  constructor(public value: number) {}
  getValue() {
    return this.value;
  }
}

class Add implements IValue {
  constructor(public value1: IValue, public value2: IValue) {}
  getValue() {
    return this.value1.getValue() + this.value2.getValue();
  }
}

class Mul implements IValue {
  constructor(public value1: IValue, public value2: IValue) {}
  getValue() {
    return this.value1.getValue() * this.value2.getValue();
  }
}

const obj = new Mul(new Add(new Value(1), new Value(2)), new Value(3));

console.log(obj.getValue());
